File: src/moleminer/registry.py
"""Source registry for discovering and managing search sources."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from moleminer.sources.base import BaseSource


class SourceRegistry:
    """Registry of all available search sources."""

    def __init__(self) -> None:
        self._sources: dict[str, type[BaseSource]] = {}

    def register(self, source_cls: type[BaseSource]) -> None:
        self._sources[source_cls.name] = source_cls

    def get_source(self, name: str) -> BaseSource:
        if name not in self._sources:
            raise KeyError(f"Source '{name}' not registered")
        return self._sources[name]()

    def get_enabled_sources(self) -> list[BaseSource]:
        return [cls() for cls in self._sources.values() if cls().enabled()]

    def list_sources(self) -> list[str]:
        return list(self._sources.keys())

File: ts/src/registry.ts
/**
 * Source registry for discovering and managing search sources.
 */

import type { Config } from './config.js';
import type { BaseSource } from './sources/base.js';

export class SourceRegistry {
  private sources = new Map<string, BaseSource>();

  /** Register a source instance. */
  register(source: BaseSource): void {
    this.sources.set(source.name, source);
  }

  /** Get a registered source by name. Throws if not found. */
  getSource(name: string): BaseSource {
    const source = this.sources.get(name);
    if (!source) {
      throw new Error(`Source '${name}' not registered`);
    }
    return source;
  }

  /** Return all sources that are currently enabled given the config. */
  getEnabledSources(config: Config): BaseSource[] {
    const result: BaseSource[] = [];
    for (const source of this.sources.values()) {
      if (source.enabled(config)) {
        result.push(source);
      }
    }
    return result;
  }

  /** List all registered source names. */
  listSources(): string[] {
    return [...this.sources.keys()];
  }
}
